Add CRD for VirtualGateway APIs for v1beta2

rajal-amzn opened the pull request:
*Issue #, if available:*
[aws-app-mesh-roadmap #37](https://github.com/aws/aws-app-mesh-roadmap/issues/37)

Description of changes:
Adding VirtualGateway CRD API version for v1beta2(based on designs #150)

Note for reviewer:
The PR is separated into three commits, only the 2nd one needs to be reviewed while the other two are auto-generated.

"auto-generate changes for adding VirtualGateway API" is auto-generated by
"kubebuilder create api --group appmesh --version v1beta2 --kind VirtualGateway"
"Separate out common fields.." contains real manual changes where I have moved out the Virtual Node types to Virtual Node class and defined new ones for Virtual Gateway since we need flexibility to evolve both Virtual Node and Virtual Gateway separately
"Auto-generate CRD APIs for Virtual Gateway" is auto-generated by "make manifests && make generate"
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

M00nF1sh commented on apis/appmesh/v1beta2/virtualgateway_types.go:
```diff
@@ -17,25 +17,262 @@ limitations under the License.
+type VirtualGatewayPortProtocol string
+
+// VirtualGatewayPortMapping refers to https://docs.aws.amazon.com/app-mesh/latest/APIReference/API_PortMapping.html
+type VirtualGatewayPortMapping struct {
```
did we introduced different structure for these(e.g. healthCheckPolicy/PortMapping) in the SDK JSON?

The types here should be closely mapping to the types in SDK SON. (except use same case for Acronym, e.g. Tls->TLS). So that we can write tools to generate it in the future.

rajal-amzn replied:
Yes. We have introduced separate structures for Virtual Gateway and Virtual Node in the AWS SDK. Refer [here](https://raw.githubusercontent.com/aws/aws-app-mesh-roadmap/master/appmesh-preview/service-model.json) for details.

karanvasnani commented on apis/appmesh/v1beta2/types.go:
```diff
@@ -226,3 +83,11 @@ type MeshReference struct {
+	// Name is the name of Mesh CR
+	Name string `json:"name"`
+	// UID is the UID of Mesh CR
+	UID types.UID `json:"uid"`
```
Shouldn't specifying `Namespace` here be sufficient as VirtualGateway is a Namespace-scoped resource unlike Mesh?

rajal-amzn replied:
Yes thats right. This has to be namespace-scoped. Will change it.

karanvasnani commented on apis/appmesh/v1beta2/types.go:
```diff
@@ -226,3 +83,11 @@ type MeshReference struct {
+
+// VirtualGatewayReference holds a reference to VirtualGateway.appmesh.k8s.aws
+type VirtualGatewayReference struct {
+	// Name is the name of Mesh CR
```
nit: VirtualGateway instead of Mesh here and below.

rajal-amzn replied:
Copy-paste error. :) Will fix it.

karanvasnani commented on apis/appmesh/v1beta2/virtualgateway_types.go:
```diff
@@ -17,25 +17,262 @@ limitations under the License.
+	// If unspecified, defaults to be 10
+	// +kubebuilder:validation:Minimum=2
+	// +kubebuilder:validation:Maximum=10
+	// +optional
+	HealthyThreshold *int64 `json:"healthyThreshold,omitempty"`
+	// The time period in milliseconds between each health check execution.
+	// If unspecified, defaults to be 30000
+	// +kubebuilder:validation:Minimum=5000
+	// +kubebuilder:validation:Maximum=300000
+	// +optional
+	IntervalMillis *int64 `json:"intervalMillis,omitempty"`
+	// The destination path for the health check request.
+	// This value is only used if the specified protocol is http or http2. For any other protocol, this value is ignored.
+	// +optional
+	Path *string `json:"path,omitempty"`
+	// The destination port for the health check request.
+	// If unspecified, defaults to be same as port defined in the PortMapping for the listener.
+	// +optional
+	Port *PortNumber `json:"port,omitempty"`
+	// The protocol for the health check request
+	// If unspecified, defaults to be same as protocol defined in the PortMapping for the listener.
+	// +optional
+	Protocol *VirtualGatewayPortProtocol `json:"protocol,omitempty"`
+	// The amount of time to wait when receiving a response from the health check, in milliseconds.
+	// If unspecified, defaults to be 5000
+	// +kubebuilder:validation:Minimum=2000
+	// +kubebuilder:validation:Maximum=60000
+	// +optional
+	TimeoutMillis *int64 `json:"timeoutMillis,omitempty"`
+	// The number of consecutive failed health checks that must occur before declaring a virtual Gateway unhealthy.
+	// If unspecified, defaults to be 2
+	// +kubebuilder:validation:Minimum=2
+	// +kubebuilder:validation:Maximum=10
+	// +optional
+	UnhealthyThreshold *int64 `json:"unhealthyThreshold,omitempty"`
```
From the model [here](https://raw.githubusercontent.com/aws/aws-app-mesh-roadmap/master/appmesh-preview/service-model.json) following should be required fields:
```
"HealthCheckPolicy": {
      "type": "structure",
      "required": [
        "healthyThreshold",
        "intervalMillis",
        "protocol",
        "timeoutMillis",
        "unhealthyThreshold"
      ]

```

rajal-amzn replied:
These values have defaults if not specified in Virtual Node. Hence I followed the same convention.

M00nF1sh commented:
general rules: 
1. type/field always stick to SDK json with one exception, use same case for abbreviation like TLS/ACM/ARN
2. if a field is optional in sdk, make it optional in crd as well, and don't use mutating webhook to set default value.
3. if a field is required in sdk, but have sensible default in k8s context, we can make it optional, but use mutating webhook to set default value.
4. if a field is required in sdk, but don't have sensible default in k8s context, we make it required.

rajal-amzn commented:
Fixed the naming a

> some general rules: :D (so we need to change virtualNodeHealthCheckPolicy/tls/etc...)
> 
> 1. type/field always stick to SDK json with one exception, use same case for abbreviation like TLS/ACM/ARN
> 2. if a field is optional in sdk, make it optional in crd as well, and don't use mutating webhook to set default value.
> 3. if a field is required in sdk, but have sensible default in k8s context, we can make it optional, but use mutating webhook to set default value.
> 4. if a field is required in sdk, but don't have sensible default in k8s context, we make it required.

Thanks. I have fixed the naming in the latest revision. So, now the VirtualNode related APIs are just moved from types.go to virtualnode_types.go without any changes. Also validated the names of VirtualGateway types to match with the SDK.
